File: test-suite-java/src/test/java/io/micronaut/interceptor/test/micronautapi/CompiledInterceptor.java
package io.micronaut.interceptor.test.micronautapi;

import io.micronaut.aop.MethodInvocationContext;
import io.micronaut.interceptor.MicronautConstructorInvocationContext;
import io.micronaut.interceptor.MicronautInvocationContext;
import io.micronaut.interceptor.MicronautMethodInvocationContext;
import jakarta.annotation.PostConstruct;
import jakarta.interceptor.AroundConstruct;
import jakarta.interceptor.AroundInvoke;
import jakarta.interceptor.Interceptor;
import jakarta.interceptor.InvocationContext;

/**
 * Reads everything it needs through the Micronaut view of the context, and never calls {@code getMethod()},
 * {@code getConstructor()} or {@code getInterceptorBindings()}.
 */
@Interceptor
@Compiled(region = "users")
public class CompiledInterceptor {

    @AroundConstruct
    public void construct(InvocationContext context) throws Exception {
        if (context instanceof MicronautConstructorInvocationContext micronaut) {
            Recorded.VALUES.add(micronaut.getInterceptorKind() + " "
                + micronaut.getBeanConstructor().getDeclaringBeanType().getSimpleName()
                + "(" + micronaut.getBeanConstructor().getArguments().length + ")"
                // the same constructor reached through the whole Micronaut construction
                + " same=" + (micronaut.getMicronautInvocation().getConstructor()
                    == micronaut.getBeanConstructor()));
        }
        context.proceed();
    }

    @PostConstruct
    public void created(InvocationContext context) throws Exception {
        record(context);
        context.proceed();
    }

    @AroundInvoke
    public Object invoke(InvocationContext context) throws Exception {
        record(context);
        return context.proceed();
    }

    private static void record(InvocationContext context) {
        if (context instanceof MicronautMethodInvocationContext micronaut) {
            Recorded.VALUES.add(micronaut.getInterceptorKind() + " "
                + micronaut.getExecutableMethod().getDeclaringType().getSimpleName() + "."
                + micronaut.getExecutableMethod().getMethodName()
                // the members of the binding, read without building the annotation
                + " region=" + region(micronaut)
                // the whole Micronaut invocation, for what the other accessors do not reach
                + named(micronaut.getMicronautInvocation()));
        }
    }

    /** An argument read by name, which the specification's own accessors have no way to do. */
    private static String named(MethodInvocationContext<Object, Object> invocation) {
        if (invocation.getParameters().isEmpty()) {
            return "";
        }
        return " parameters=" + invocation.getParameters().keySet();
    }

    private static String region(MicronautInvocationContext context) {
        return context.getAnnotationMetadata().stringValue(Compiled.class, "region").orElse("none");
    }
}

File: test-suite-kotlin/src/test/kotlin/io/micronaut/interceptor/test/CompiledInterceptor.kt
package io.micronaut.interceptor.test

import io.micronaut.interceptor.MicronautConstructorInvocationContext
import io.micronaut.interceptor.MicronautMethodInvocationContext
import jakarta.annotation.PostConstruct
import jakarta.interceptor.AroundConstruct
import jakarta.interceptor.AroundInvoke
import jakarta.interceptor.Interceptor
import jakarta.interceptor.InvocationContext

/**
 * Reads the interception as Micronaut compiled it, and never calls getMethod(), getConstructor() or
 * getInterceptorBindings().
 */
@Interceptor
@Compiled(region = "users")
open class CompiledInterceptor {

    @AroundConstruct
    open fun construct(context: InvocationContext) {
        if (context is MicronautConstructorInvocationContext) {
            val constructor = context.beanConstructor
            // the same constructor, reached through the whole Micronaut construction
            val same = context.micronautInvocation.constructor === constructor
            Calls.recorded += "${context.interceptorKind} " +
                "${constructor.declaringBeanType.simpleName}(${constructor.arguments.size}) same=$same"
        }
        context.proceed()
    }

    @PostConstruct
    open fun created(context: InvocationContext) {
        record(context)
        context.proceed()
    }

    @AroundInvoke
    open fun invoke(context: InvocationContext): Any? {
        record(context)
        return context.proceed()
    }

    private fun record(context: InvocationContext) {
        if (context is MicronautMethodInvocationContext) {
            val method = context.executableMethod
            // the members of the binding, read without building the annotation
            val region = context.annotationMetadata.stringValue(Compiled::class.java, "region").orElse("none")
            // an argument by name, which the specification's own accessors have no way to reach
            val names = context.micronautInvocation.parameters.keys
            val parameters = if (names.isEmpty()) "" else " parameters=$names"
            Calls.recorded += "${context.interceptorKind} " +
                "${method.declaringType.simpleName}.${method.methodName} region=$region$parameters"
        }
    }
}
